src/drivers/uavcan/node_info.cpp

```diff
@@ -0,0 +1,335 @@
+
+			if (is_registering_info) {
+				populateDeviceInfoFields(_device_informations[i], *info);
+
+			} else {
+				_device_informations[i].device_id = device_id;
+				_device_informations[i].capability = capability;
+			}
+
+			return;
+		}
+	}
+
+	// Case 4: extend array and add entry at the end
+	if (extendDeviceInformationsArray()) {
+		_device_informations[_device_informations_size - 1] = DeviceInformation();
+		_device_informations[_device_informations_size - 1].node_id = node_id;
+
+		if (is_registering_info) {
+			populateDeviceInfoFields(_device_informations[_device_informations_size - 1], *info);
+
+		} else {
+			_device_informations[_device_informations_size - 1].device_id = device_id;
+			_device_informations[_device_informations_size - 1].capability = capability;
+		}
+
+	} else {
+		PX4_DEBUG("Failed to extend device informations array for %s",
+			  is_registering_info ? "node info" : "capability");
+	}
+}
+
+void NodeInfoPublisher::registerDeviceCapability(uint8_t node_id, uint32_t device_id, DeviceCapability capability)
+{
+	registerDevice(node_id, nullptr, device_id, capability);
+}
+
+void NodeInfoPublisher::publishDeviceInformationImmediate(size_t device_index)
+{
+	if (device_index >= _device_informations_size) {
+		return;
+	}
+
+	const auto &device_info = _device_informations[device_index];
+
+	const uint64_t now = hrt_absolute_time();
+
+	// Rate limit immediate publishing to prevent spam
+	if (now < _last_device_info_publish + DEVICE_INFO_PUBLISH_RATE_LIMIT_US) {
+		PX4_DEBUG("Immediate publish for node %d rate limited",
+			  device_info.node_id);
+		return;
+	}
+
+	_last_device_info_publish = now;
+	publishSingleDeviceInformation(device_info);
+}
+
+void NodeInfoPublisher::publishDeviceInformationPeriodic()
+{
+	// Using round-robin approach to publish one device info per timer event
+	if (_device_informations_size == 0) {
+		return;
+	}
+
+	size_t devices_checked = 0;
+
+	while (devices_checked < _device_informations_size) {
+		if (_next_device_to_publish >= _device_informations_size) {
+			_next_device_to_publish = 0;
+		}
+
+		const auto &device_info = _device_informations[_next_device_to_publish];
+
+		if (device_info.has_node_info && device_info.capability != DeviceCapability::NONE) {
+			publishSingleDeviceInformation(device_info);
+			_next_device_to_publish++;
+			return;
+		}
+
+		_next_device_to_publish++;
+		devices_checked++;
+	}
+
+	PX4_DEBUG("No devices ready for periodic publishing");
+}
+
+void NodeInfoPublisher::publishSingleDeviceInformation(const DeviceInformation &device_info)
+{
+	const uint64_t now = hrt_absolute_time();
+
+	device_information_s msg{};
+	msg.timestamp = now;
+	msg.device_type = static_cast<uint8_t>(device_info.capability);
+	msg.device_id = device_info.device_id;
+
+	// Copy pre-populated fields directly from the struct
+	// Copy strings using memcpy and ensure null termination
+	memcpy(msg.model_name, device_info.model_name, sizeof(msg.model_name));
```

Comment:
In case someone decides to change one of the array sizes, you can add this to ensure they are kept in sync: `static_assert(sizeof(msg.model_name) == sizeof(device_info.model_name), "Array size mismatch");`